File: src/js/ts/js/src/AbstractSoundData.ts
class AbstractSoundData {

    protected id: string;
    protected isMuted: boolean;
    protected currentStatus: SoundStatus;
    protected currentVolume: number;
    protected mutedGroupList: string[];
    protected pausedGroupList: string[];
    protected effectCommandList: AbstractEffectCommand[];

    protected parentGroupId: string;

    constructor(id: string) {
        this.id = id;
        this.isMuted = false;
        this.currentStatus = SoundStatus.PLAYING;
        this.currentVolume = 1.0;

        this.mutedGroupList = [];
        this.pausedGroupList = [];
        this.effectCommandList = [];
    }

    public getId(): string {
        return this.id;
    }

    // tslint:disable-next-line:no-empty
    public writeRawPCM(pcmBufferArray: number[][]): void {}

    // tslint:disable-next-line:no-empty
    public pause(): void {}

    // tslint:disable-next-line:no-empty
    public resume(): void {}

    // tslint:disable-next-line:no-empty
    public stop(): void {}

    // tslint:disable-next-line:no-empty
    public mute(): void {}

    // tslint:disable-next-line:no-empty
    public unmute(): void {}

    public getVolume(): number {
        return this.currentVolume;
    }

    public setVolume(volume: number): void {
        this.currentVolume = volume;
    }

    public getParent(): string {
        return this.parentGroupId;
    }

    public setParent(parentGroupId: string) {
        this.parentGroupId = parentGroupId;
    }

    // tslint:disable-next-line:no-empty
    public _pauseByGroup(soundGroupId: string) {}

    // tslint:disable-next-line:no-empty
    public _resumeByGroup(soundGroupId: string) {}

    // tslint:disable-next-line:no-empty
    public _muteByGroup(soundGroupId: string) {}

    // tslint:disable-next-line:no-empty
    public _unmuteByGroup(soundGroupId: string) {}
}

File: src/c/AbstractSoundData.cpp
/**
*  Harmonia
*
*  Copyright (c) 2018 Hiiragi
*
*  This software is released under the MIT License.
*  http://opensource.org/licenses/mit-license.php
*/

#include "AbstractSoundData.h"
#include "SoundStatus.h"

#include <algorithm>
#include <string>

AbstractSoundData::AbstractSoundData(std::string id)
{
	_id = id;
	_isMuted = false;
	_currentStatus = SoundStatus::PLAYING;
	_currentVolume = 1.0f;


	_mutedGroupList = new std::list<std::string>();
	_pausedGroupList = new std::list<std::string>();

	_effectCommandList = new std::list<AbstractEffectCommand*>();
}


AbstractSoundData::~AbstractSoundData()
{
	delete _mutedGroupList;
	delete _pausedGroupList;

	size_t size = _effectCommandList->size();

	if (size > 0)
	{
		std::for_each(_effectCommandList->cbegin(), _effectCommandList->cend(), [](AbstractEffectCommand* command) {
			delete command;
		});
	}

	delete _effectCommandList;
}

void AbstractSoundData::write_raw_pcm(int* combinedBuffer, int size)
{
}

void AbstractSoundData::pause()
{
}

void AbstractSoundData::resume()
{
}

void AbstractSoundData::stop()
{
}

void AbstractSoundData::mute()
{
}

void AbstractSoundData::unmute()
{
}

std::string AbstractSoundData::get_id()
{
	return _id;
}

float AbstractSoundData::get_volume()
{
	return _currentVolume;
}

void AbstractSoundData::set_volume(float volume)
{
	if (volume > 1)
	{
		_currentVolume = 1;
	}
	else if (volume < 0)
	{
		_currentVolume = 0;
	}
	else
	{
		_currentVolume = volume;
	}
}

SoundStatus AbstractSoundData::get_status()
{
	return _currentStatus;
}

std::string AbstractSoundData::get_parent()
{
	return _parentGroupId;
}
void AbstractSoundData::_setParent(std::string parentGroupId)
{
	_parentGroupId = parentGroupId;
}

void AbstractSoundData::_pause_by_group(std::string soundGroupId)
{
}

void AbstractSoundData::_resume_by_group(std::string soundGroupId)
{
}

void AbstractSoundData::_mute_by_group(std::string soundGroupId)
{
}

void AbstractSoundData::_unmute_by_group(std::string soundGroupId)
{
}
